team-validator.js

```diff
@@ -892,24 +892,15 @@ class Validator {
 						// Defog and Whirlpool can't be transferred together
 						if (dex.gen >= 5 && moveid in {'defog':1, 'whirlpool':1} && learnedGen <= 4) blockedHM = true;
 					}
-					if (learned.substr(0, 2) in {'4L':1, '5L':1, '6L':1, '7L':1}) {
-						// gen 4-7 level-up moves
-						if (level >= parseInt(learned.substr(2)) || learnedGen === 7 && dex.gen >= 7) {
-							// we're past the required level to learn it
-							return false;
```

Comment:
The correct solution appears to be to change this to set `sourcesBefore`

Reply:
The `if (learned.charAt(1) in {L:1, M:1, T:1})` block already knows about setting `sourcesBefore`, so I don't need to repeat it. By only handling underlevelled Gen4-6 moves I even avoid having to do a Gen7 check.

Reply:
Oh, I understand now.

Reply:
I WAS RIGHT! THE EXACTLY CORRECT SOLUTION WOULD HAVE BEEN TO SET `sourcesBefore`

Reply:
The problem here is that apparently you don't understand enough of how the validator works. The validator is some pretty precise code.

The code is designed to look for ways you can learn the move. As a shortcut, you can `return false` to mean "no restrictions on learning the move".

Otherwise: You need to add ways to learn the move to `sources` or `sourcesBefore`. By not doing that, your code is saying that this move can't be learned at all.

The old code set `learned` to `Eany` and then fell through to let the egg move validator set `sources`. Your new code replaced the `if` with an `else if`, stopping it from falling through, which caused the LC bugs people are reporting today.

Reply:
Sorry about that, that case was too subtle for me to pick up on.

But you still don't need to handle Gen 7 moves at all, so it could have looked something like this:

    if (learned.substr(0, 2) in {'4L':1, '5L':1, '6L':1} && level < parseInt(learned.substr(2))) {
        if (template.gender && template.gender !== 'F') continue;
        // available as egg move
        learned = learnedGen + 'Eany';
        limitedEgg = false;
    }

Reply:
What _I_ don't understand is why there isn't also checking for gen 1-3 moves?